api.md

```diff
@@ -390,6 +391,242 @@ if you try to upload another file other than an image:
 Or an image size that's too large,
 you'll get an `413 Request Entity Too Large` error.
 
+# 5. Returning meaningful errors
 
+If something *fails*, the person using the API
+will have a `JSON` body stating 
+`"Error uploading file #26"` returned.
+This type of information is not particularly *useful*.
+We don't know exactly *why* it failed, 
+we just know it did.
 
+In order to fix this, 
+let's make some changes to tell the person
+a more meaningful message so they know *what went wrong*.
+We want the person to know whether any of these scenarios:
+- reading the `path` from the `image`.
+- creating the `CID` from the file contents.
+- getting the `file extension`.
+- uploading the file to `S3`.
 
+Before making any changes, let's add a few tests to cover these.
+First, let's create an empty file, which will be used in our tests.
+In `priv/static/images`, create an empty file called `empty`.
+
+After this,
+open `test/app_web/api_test.exs` and add this.
+
+```elixir
+  import Mock
+
+  @empty_file %{
+    "" => %Plug.Upload{
+      content_type: "image/something",
+      filename: "empty",
+      path: [:code.priv_dir(:app), "static", "images", "empty"] |> Path.join()
+    }
+  }
+
+  test "empty file should return appropriate error", %{conn: conn} do
+    conn = post(conn, ~p"/api/images", @empty_file)
+
+    assert Map.get(Jason.decode!(response(conn, 400)), "errors") == %{
+             "detail" => "Error uploading file. The file extension and contents are invalid."
+           }
+  end
+
+  test "empty image (meaning it has a valid content type) should return an error", %{conn: conn} do
+    conn = post(conn, ~p"/api/images", @empty_image)
+
+    assert Map.get(Jason.decode!(response(conn, 400)), "errors") == %{
+             "detail" => "Error uploading file. Failure creating the CID filename."
+           }
+  end
+
+  test "valid file but the upload to S3 failed. It should return an error.", %{conn: conn} do
+
+    with_mock ExAws, [request: fn(_input) -> {:error, :failure} end] do
+      conn = post(conn, ~p"/api/images", @valid_image_attrs)
+
+      assert Map.get(Jason.decode!(response(conn, 400)), "errors") == %{
+               "detail" => "Error uploading file. There was an error uploading the file to S3."
+             }
+    end
+  end
+```
+
+> **Note** 
+>
+> We are using the [`mock` ](https://github.com/jjh42/mock)
+> library so we are able to correctly test 
+> whenever `S3` is down.
+> Although using mocks *can lead to false positives*,
+> here's acceptable because we're simulating a third-party service to be down.
+
+
+We've added a test for each scenario 
```

Comment:
This is not a "test"; It's a _`mock`_. 
![image](https://github.com/dwyl/imgup/assets/194400/a6a939e2-0106-4162-b440-145b9d76c805)

It might _look_ like a "test" to the untrained eye, 
but to the person that _cares_ it's **_much_ worse** than having nothing at all.

Reply:
Given that `mocks` were deleted, this is fixed 👌